tsfc/coffee_mode.py

```diff
@@ -0,0 +1,294 @@
+
+
+def optimise_expressions(expressions, argument_multiindices):
+    """Perform loop optimisations on GEM DAGs
+
+    :arg expressions: list of GEM DAGs
+    :arg argument_multiindices: tuple of argument multiindices,
+                                one multiindex for each argumen
+
+    :returns: list of optimised GEM DAGs
+    """
+    # Propagate Failure nodes
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return expressions
+
+    def classify(argument_indices, expression):
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, (Indexed, Conditional)):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+
+    argument_indices = tuple(itertools.chain(*argument_multiindices))
+    # Apply argument factorisation unconditionally
+    classifier = partial(classify, set(argument_indices))
+    monomial_sums = collect_monomials(expressions, classifier)
+    monomial_sums = [optimise_monomial_sum(ms, argument_indices) for ms in monomial_sums]
+    return list(map(monomial_sum_to_expression, monomial_sums))
+
+
+def index_extent(factor, argument_indices):
+    """Compute the product of the extents of argument indices of a GEM expression
+
+    :arg factor: GEM expression
+    :arg argument_indices: set of argument indices
+
+    :returns: product of extents of argument indices
+    """
+    return numpy.product([i.extent for i in set(factor.free_indices).intersection(argument_indices)])
+
+
+def monomial_sum_to_expression(monomial_sum):
+    """Convert a monomial sum to a GEM expression.
+
+    :arg monomial_sum: :class:`MonomialSum` object
+
+    :returns: GEM expression
+    """
+    indexsums = []  # The result is summation of indexsums
+    # Group monomials according to their sum indices
+    groups = groupby(monomial_sum, key=lambda m: frozenset(m.sum_indices))
+    # Create IndexSum's from each monomial group
+    for _, monomials in groups:
+        # Pick sum indices from the first monomial
+        sum_indices = monomials[0].sum_indices
+        # Create one product for each monomial
+        products = [make_product(monomial.atomics + (monomial.rest,)) for monomial in monomials]
+        indexsums.append(IndexSum(make_sum(products), sum_indices))
+
+    return make_sum(indexsums)
+
+
+def find_optimal_atomics(monomial_sum, argument_indices):
+    """Find optimal atomic common subexpressions, which produce least number of
+    terms in the resultant IndexSum when factorised.
+
+    :arg monomial_sum: A :class:`MonomialSum` object, or a iterable collection
+                       of monomials, all of the monomials should have the
+                       same sum indices
+    :arg argument_indices: tuple of argument indices
+
+    :returns: list of atomic GEM expressions
```

Comment:
I think this is a lie.  In one case you return a 2-tuple of empty tuples.  In another, you return a 2-tuple of a single atomic and an empty tuple, in the third case you return an n-tuple of atomics, it looks like.

The return value of this function is only used inside `factor_atomics` which seems to replicate some of the unpicking logic to do different things depending on how long the list of optimal atomics is.  This seems like we could think a little more about how they fit together and clean up.